keps/sig-apps/20190226-maxunavailable-for-statefulsets.md

```diff
@@ -0,0 +1,226 @@
+        Partition *int32 `json:"partition,omitempty" protobuf:"varint,1,opt,name=partition"`
+
+	// NOTE THIS IS THE NEW FIELD BEING PROPOSED
+	// The maximum number of pods that can be unavailable during the update.
+        // Value can be an absolute number (ex: 5) or a percentage of desired pods (ex: 10%).
+        // Absolute number is calculated from percentage by rounding down.
+        // Defaults to 1.
+        // +optional
+        MaxUnavailable *intstr.IntOrString `json:"maxUnavailable,omitempty" protobuf:"bytes,2,opt,name=maxUnavailable"`
+	
+	...
+}
+```
+
+- By Default, if maxUnavailable is not specified, its value will be assumed to be 1 and StatefulSets will follow their old behavior. This
+  will also help while upgrading from a release which doesnt support maxUnavailable to a release which supports this field.
+- If maxUnavailable is specified, it cannot be greater than total number of replicas.
+- If maxUnavailable is specified and partition is also specified, MaxUnavailable cannot be greater than `replicas-partition`
+- podManagementPolicy = OrderedReady
+  If a partition is specified, maxUnavailable will only apply to all the pods which are staged by the partition. Which means all Pods with 
+an ordinal that is greater than or equal to the partition will be updated when the StatefulSet’s .spec.template is updated. Lets say total 
+replicas is 5 and partition is set to 2 and maxUnavailable is set to 2. If the image is changed in this scenario, pods with ordinal 4 and 3 will go
+down at the same time(because of maxUnavailable), once they are running and ready, pods with ordinal 2 will go down. Pods with ordinal 0
+and 1 will remain untouched due the partition.
+- podManagementPolicy = [Parallel](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#parallel-pod-management)
+  here maxUnavailable does not make sense since in this policy, pods are launched or terminated in parallel and do not wait for pods to become
+  running and ready. So we will never know when two terminate the next batch of maxUnavailable pods.
+
+
+#### Implementation
+
+https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/statefulset/stateful_set_control.go#L504
+```go
+...
+	podsDeleted := 0
+	// we terminate the Pod with the largest ordinal that does not match the update revision.
+	for target := len(replicas) - 1; target >= updateMin; target-- {
+
+		// delete the Pod if it is not already terminating and does not match the update revision.
+		if getPodRevision(replicas[target]) != updateRevision.Name && !isTerminating(replicas[target]) {
+			klog.V(2).Infof("StatefulSet %s/%s terminating Pod %s for update",
+				set.Namespace,
+				set.Name,
+				replicas[target].Name)
+			err := ssc.podControl.DeleteStatefulPod(set, replicas[target])
+			status.CurrentReplicas--
+
+			// NEW CODE HERE
+			if podsDeleted < set.Spec.UpdateStrategy.RollingUpdate.MaxUnavailable {
+				podsDeleted ++;
+				continue;
+			}
```

Comment:
We should describe how this feature can be implemented and how it interacts with partition and pod management policy, rather than a code block.

Reply:
I think the implementation cannot be concretely defined, untill we agree on the semantics. I will update the doc with the sematics in case of partition . Also as you rightly mentioned earlier, the maxUnavailable has nothing to do with pod management policy. So i will update the docs accordingly. Once we have agreed on the semantics i will update the implementation section as well.